[clang] Allow `ConditionalOperator` fast-math flags to be  overridden by `pragma float_control`

echesakov opened the pull request:
Currently, the fast-math flags set on `select` or `phi` instruction
emitted by CodeGen when visiting `ConditionalOperator` take into account
only global `FPOptions` and ignore `pragma float_control`.

This involves storing `FPOptionsOverride` in trailing storage of
`ConditionalOperator` and storing `CurFPOptionsOverride` when creating
an AST node.

Fixes #84648

echesakov commented:
Ping

arsenm commented:
Can you reword the description to have less wrapping

arsenm commented on clang/test/CodeGen/conditional-operator.c:
```diff
@@ -0,0 +1,36 @@
+#pragma float_control(precise, on)
+  return c ? t : f;
+}
+
```
Test some vector and complex cases

echesakov replied:
For "vector case" do you mean something along the lines of

```c
typedef float float2 __attribute__((ext_vector_type(2)));
typedef int int2 __attribute__((ext_vector_type(2)));

float2 foo(int2 c, float2 t, float2 f) {
#pragma float_control(precise, off)
    return c ? t : f;
}
```

In this case, the expression would be codegen-d to a sequence of bitwise operations and these are not affected by fast-math flags

```llvm
; Function Attrs: noinline nounwind optnone
define <2 x float> @foo(<2 x i32> noundef %c, <2 x float> noundef %t, <2 x float> noundef %f) #0 {
entry:
  %c.addr = alloca <2 x i32>, align 8
  %t.addr = alloca <2 x float>, align 8
  %f.addr = alloca <2 x float>, align 8
  store <2 x i32> %c, ptr %c.addr, align 8
  store <2 x float> %t, ptr %t.addr, align 8
  store <2 x float> %f, ptr %f.addr, align 8
  %0 = load <2 x i32>, ptr %c.addr, align 8
  %1 = load <2 x float>, ptr %t.addr, align 8
  %2 = load <2 x float>, ptr %f.addr, align 8
  %3 = icmp slt <2 x i32> %0, zeroinitializer
  %sext = sext <2 x i1> %3 to <2 x i32>
  %4 = xor <2 x i32> %sext, <i32 -1, i32 -1>
  %5 = bitcast <2 x float> %2 to <2 x i32>
  %6 = bitcast <2 x float> %1 to <2 x i32>
  %7 = and <2 x i32> %5, %4
  %8 = and <2 x i32> %6, %sext
  %cond = or <2 x i32> %7, %8
  %9 = bitcast <2 x i32> %cond to <2 x float>
  ret <2 x float> %9
}
```

Not really sure what I will be testing here?

arsenm commented on clang/test/CodeGen/conditional-operator.c:
```diff
@@ -0,0 +1,36 @@
+// RUN: %clang_cc1 -disable-llvm-passes -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -disable-llvm-passes -emit-llvm -menable-no-infs -fapprox-func\
+// RUN: -funsafe-math-optimizations -fno-signed-zeros -mreassociate -freciprocal-math\
+// RUN: -ffp-contract=fast -ffast-math %s -o - | FileCheck %s
```
Redundantly using -ffast-math and all the individual flags?

echesakov replied:
These flags correspond to how the clang driver `-ffast-math` are expanded to the individual flags passed to `-cc1`.

Several tests (e.g. `clang/test/Parser/fp-floatcontrol-syntax.cpp` and `clang/test/CodeGen/math-errno.c`) use similar command lines for testing fast-path behavior and I was trying to be consistent with them.

zahiraam commented:
Not very clear to me what the issue is. Can you please explain this in the description. I think here the `#pragma float_control(precise, on)` should be taking precedence over the `-ffast-math` flag and it's not. Right?

echesakov commented:
> Not very clear to me what the issue is. Can you please explain this in the description. I think here the `#pragma float_control(precise, on)` should be taking precedence over the `-ffast-math` flag and it's not. Right?

@zahiraam Correct, this is not happening right now and flags on codegen-d `select`/`phi` are not taking into account the pragma.

echesakov committed:
[clang] Allow `ConditionalOperator` fast-math flags to be overridden by `pragma float_control`

Currently, the fast-math flags set on `select` or `phi` instruction
emitted by CodeGen when visiting `ConditionalOperator` take into account
only global `FPOptions` and ignore `pragma float_control`.

This involves storing `FPOptionsOverride` in trailing storage of
`ConditionalOperator` and storing `CurFPOptionsOverride` when creating
an AST node.

Fixes #84648

echesakov commented:
> Can you reword the description to have less wrapping

@arsenm @zahiraam Updated the commit/PR and added more details on the issue.

spavloff commented:
I don't think think setting FPOptions in `ConditionalOperator` is a good idea. This operator does not represent any floating-point operation so setting FP properties on it does not look reasonable.

Instead, I would like to propose you using FPOptions from CompoundStmt, you could use the same code as in this PR: https://github.com/llvm/llvm-project/pull/89617/files#diff-6ddbe6b52b91cf337ee1b7c1ce6ced7d81882f749ac36c81d12714cb298319d6R568-R573
It was added just for similar cases.

rjmccall commented:
Yeah, the conditional operator doesn't do any floating-path math itself.  If the first operand is a floating-point expression, we should always be modeling that with a float-to-boolean conversion, and the flags should go there.

andykaylor commented:
> Yeah, the conditional operator doesn't do any floating-path math itself. If the first operand is a floating-point expression, we should always be modeling that with a float-to-boolean conversion, and the flags should go there (if they're necessary — I didn't think comparisons were flag-sensitive).

Any instruction that returns a floating-point value can have fast-math flags attached to it, and because we need to check the fast-math flags on input operands, the select instruction generated by the conditional operator needs the flags set when fast-math is enabled. That's happening in the general case, but it isn't currently being modified correctly in the presence of pragmas.

https://godbolt.org/z/91zPx5ha4

spavloff commented:
> Any instruction that returns a floating-point value can have fast-math flags attached to it, and because we need to check the fast-math flags on input operands, the select instruction generated by the conditional operator needs the flags set when fast-math is enabled. That's happening in the general case, but it isn't currently being modified correctly in the presence of pragmas.

Clang should not need similar workaround. FP pragmas have block scope, so their effect should be recorded in CompoundStmt. When entering this scope IRBuilder must be tuned according to FPOptions, stored in CompoundStmt. Now the code that tunes IRBuilder is absent, if we add it, it should implement the required behavior.

andykaylor commented:
> Are you suggesting that loads need to have fast-math flags attached to them? Because this sounds like a bad representation in IR.

What I'm primarily suggesting is that we should be handling conditional operators consistently in the presence of pragmas. Whether that can be done with CompoundStmt or requires some other intervention, I don't know.

In the example I linked above, clang is attaching fast-math flags to a phi instruction when fast-math flags are used on the command line, which causes the optimizer to keep the flags and move them to a select instruction when the call graph is simplified. However, when a pragma is used to disable fast-math locally, clang still puts the fast-math flags on the phi instruction, which is a bug.

The question of whether we should be putting fast-math flags on load instructions has come up before (see https://github.com/llvm/llvm-project/issues/51601), and we resisted doing it. The current solution for that case is to use function attributes to deduce that fast-math rules apply throughout the function. That works, but it is more conservative than we'd like in cases where a function with fast-math enabled is inlined into a function with fast-math disabled. I don't have a solution to that other than attaching fast-math flags to the load instruction. I'm just noting it as a problem. It's certainly better to have the conservative behavior than to apply fast-math where it shouldn't be applied (which can happen currently with the conditional operator).

rjmccall commented:
I don't understand what fast-math flags are supposed to mean for loads, phis, and selects.  These are not arithmetic operations; they just propagate values.  If you're trying to implement some kind of rule where fast-math analysis should not pass across certain kinds of abstraction, you need to be holistically designing the IR around that, not gradually accreting flags onto every opaque data instruction.

andykaylor commented:
@rjmccall I understand your point, and I think you're raising a good question. Let's walk through an example that illustrates why we currently want FMF on phis and selects and see if we can agree on an alternative way to handle it. In a comment on https://github.com/llvm/llvm-project/issues/51601, I started with this:

```
double floatingAbs(double x) {
  return (x < 0) ? -x : x;
}

```
We want to optimize that to `llvm.fabs(x)`, but we can only do that if we don't care about the sign of zero. I walked through the steps of how that happens [here](https://github.com/llvm/llvm-project/issues/51601#issuecomment-981047527), but let me jump to the optimized IR just before the replacement happens because that's sufficient for the discussion about FMF on select instructions. The optimizer reduces the IR to this:

```
define dso_local double @floatingAbs(double %0) {
  %2 = fcmp fast olt double %0, 0.000000e+00
  %3 = fneg fast double %0
  %4 = select fast i1 %2, double %3, double %0
  ret double %4
}
```
We need the `nsz` flag to turn this into `llvm.fabs(x)`. The `nsz` flag is present on the `fcmp` instruction, but that doesn't matter because `0.0` compares as equal to `-0.0` with or without the flag. We also have `nsz` on the `fneg` instruction, but again that doesn't matter because we don't hit the `fneg` instruction for `0.0` or `-0.0`. We want this sequence to produce the absolute value of `x` but the only way we can say that it does is if we don't care about the sign of zero on the `select` instruction.

If this code is called with `x == -0.0`, the `select` instruction will return -0.0. We can only replace this with `llvm.fabs(x)` if we know we are allowed to ignore the sign of zero for the whole pattern. That leaves us two choices: either we depend on a function attribute saying that we can ignore the sign of zero for the entire function, or we must have the `nsz` flag set on all instructions involved in the pattern.

As I said before, relying on the function attribute gives us correct results, but because functions can have mixed fast-math states (through either inlining or pragmas), we may lose an optimization here. To me, it seems that having fast-math flags on `select` instruction is easiest way to do this without potentially losing optimizations, and that indirectly implies that we'd like to have FMF on phis and loads. Currently, we're in a mixed state in that regard where we accept the loss of optimization if a load is involved but preserve optimization through phis and selects.

However, as I think about this it occurs to me that there is another possibility. The argument about only applies to the `nsz` flag. The `nnan` and `ninf` flags can be deduced, and the rewrite flags don't have any meaning for phis, loads, and selects. Perhaps we shouldn't be looking at the `select` instruction but instead should be looking for the `nsz` flag on the uses of the select instruction. In the trivial case I cited above, the result of the select is being returned, so looking at the function attribute is correct. If the value selected were being used in the function, we would look at the uses. If all of the uses have the `nsz` flag set, This would complicate the handling a bit, but it certainly seems to make more sense in terms of the semantics of the `nsz` flag.

Tagging @jcranmer-intel who has been working on clarifying FMF semantics.

rjmccall commented:
Oh, that's interesting. I'd been assuming this was a cross-function issue or something like that, but that's a great example of where we need more than that.  I agree that it feels like having some kind of barrier instruction is the right way to go — basically a unary operator that can have fast-math restrictions on its operand/result just like all the binary operators and intrinsics do, and then we can decide the right places to emit that.

echesakov commented:
> I made an alternative PR, which does not need changes in ConditionalOperator: https://github.com/llvm/llvm-project/pull/111654.

Thank you @spavloff , closing in favor of #111654
